replace build-a-pki.sh with Rust+rcgen, rcgen 0.13

cpu opened the pull request:
Previously we used a bash script and the `openssl` command line tooling to generate a test PKI comprising client/end entity certificates, intermediate certificates, root certificates, and CRLs for a variety of algorithms and key sizes. This branch replaces that tooling with a Rust crate using [rcgen](https://github.com/rustls/rcgen). This will be easier to maintain over time (_who likes writing Bash....?_) and helps validate and test work underway in the rcgen crate.

As much as possible the generated certificates have been matched to the important properties of the original OpenSSL test certificates. The one significant divergence in this branch is that the RSA PKI has been spit. Previously it was one chain with a client and end-entity cert using RSA 2048, an intermediate using RSA 3072, and a root using RSA 4096. In the new iteration we match the ECDSA PKIs and make three chains: one using all RSA 2048, one using all 3072, and one using all 4096. This change requires two updates to unit tests: 

1. Because the all RSA 2048 chain is smaller than the mixed RSA chain we have to adjust an expected write byte count down slightly to account for the fact we send less data on the wire with the smaller chain.
2. The expected subject common name for the RSA CA is split into three, one per size, and the encoded bytes have changed due to the longer name that now includes the key size.

cpu commented on Cargo.toml:
```diff
@@ -2,27 +2,35 @@
+  "rustls",
   # experimental post-quantum algorithm support
   "rustls-post-quantum",
+  # test certificates and keys for rustls unit tests
+  "test-ca",
```
I started out by writing a single file Cargo script like the [`update_valid_ee_certs.rs`](https://github.com/rustls/rustls-platform-verifier/blob/main/rustls-platform-verifier/src/tests/verification_real_world/update_valid_ee_certs.rs) tool in the rustls-platform-verifier repo but ended up going back to a boring-old binary crate for a couple reasons:

1. There's no `cargo fmt` or `cargo clippy` support for these single file scripts at the current time, which is a huge downside IMO.
2. Running the script requires using nightly rust and that's a bit of a bummer.

djc replied:
How about making this an `example` in one of the existing crates?

cpu replied:
I've moved the tool to `rustls/examples/internal`.

cpu commented:
> rustls / MSRV (pull_request) Failing after 13s

Looks to be from the use of the Cargo patch:
> error: failed to load source for dependency `rcgen`

est31 commented:
CI fails with:
> The package requires the Cargo feature called `workspace-inheritance`, but that feature is not stabilized in this version of Cargo (1.61.0 (a028ae4 2022-04-29)).

workspace inheritance was [stabilized in 1.64.0](https://github.com/rust-lang/cargo/pull/10859), but rustls MSRV is 1.61.0 as per https://github.com/rustls/rustls/pull/1516.

rgcen has a rolling MSRV model of 7 months support and currently the MSRV of rcgen is [1.71.1](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1711-2023-08-03).

I suppose that CI job is not attempting to build any of the dev-dependencies, but is instead to make sure that lib users can still use rustls. However, cargo still wants to build a resolved dependency graph (including Cargo.lock) and for that it touches rcgen.

The workspace inheritance might be an artifact of the PR using git dependencies: cargo does some normalization upon publication, and I *think* this also resolves all workspace inheritance and replaces it with concrete values. So maybe once we have an rcgen release and use it in this PR, this error will go away.

There might be future errors though, due to other cargo features that are not erased upon `cargo package`. Therefore, it might be a good idea to think about harmonization of rcgen and rustls MSRVs, but this is not a neccessity. We could also work on a by-case basis.

est31 commented:
@cpu thanks for this amazing PR! It's a great demo of `rcgen` and also a good demo of rcgen's `aws_lc_rs` backend.

cpu commented:
>  So maybe once we have an rcgen release and use it in this PR, this error will go away.

That's my current suspicion, but if it turns out to be wrong we can reevaluate the right fix.

ctz commented on test-ca/src/main.rs:
```diff
@@ -1,3 +1,361 @@
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::Duration;
+
+use rcgen::RsaKeySize;
+use rcgen::{
```
Merge these two imports?

ctz reviewed (approved):
Very nice!

djc commented on test-ca/Cargo.toml:
```diff
@@ -5,3 +5,5 @@ edition = "2021"
 publish = false
 
 [dependencies]
+rcgen = { version = "0.13.0", default-features = false, features = ["aws_lc_rs", "pem"] }
```
Nit: `0.13`.

cpu replied:
Fixed.

djc commented on rustls/tests/common/mod.rs:
```diff
@@ -126,6 +126,40 @@ embed_files! {
+    (RSA_3072_INTER_CERT, "rsa-3072", "inter.cert");
+    (RSA_3072_INTER_KEY, "rsa-3072", "inter.key");
+
+    (RSA_4096_CA_CERT, "rsa-4096", "ca.cert");
```
Nit: your commit message references 4097-bit key sizes twice.

cpu replied:
Fixed.

djc reviewed (commented):
Some minor nits, nice work!

cpu commented:
> How about making this an example in one of the existing crates?

I'm open to that. It seems reasonable to put it into the `examples` crate alongside `limitedclient.rs` and friends. We already take a `rcgen` dependency there for the `server_acceptor.rs` example, and I often use the test-ca certs when running the other examples.

WDYT @ctz ? Any pref?

cpu commented:
> Perhaps it could be in rustls/examples/internal?

That works for me :+1:

cpu commented:
>> Perhaps it could be in rustls/examples/internal?
>
> That works for me 👍

I moved the tooling into `rustls/examples/internal/test_ca.rs`. I also added a short README in `test-ca` pointing to the tool. I left the commit that alphabetized the Cargo workspace members even though I'm not adding a new member anymore since we might as well do that tidying for the next time a new crate is needed.

cpu commented:
Starting to prep for the rcgen release required to unblock this in https://github.com/rustls/rcgen/pull/254

cpu commented:
Potential follow-up: RiiR [bogo regen-certs](https://github.com/rustls/rustls/blob/main/bogo/regen-certs) script.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+    }
+}
+
+struct CertifiedKey {
```
Why not use `rcgen::CertifiedKey` for this?

cpu replied:
Done.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+    ];
+
+    let mut certified_keys = HashMap::with_capacity(roles.len() * sig_algs.len());
+    for role in roles {
```
If each `(role, alg)` test is pretty much independent, I don't think we need three loops here? And if that's true, do we still need the hash map?

cpu replied:
I pulled everything into one loop.

> And if that's true, do we still need the hash map?

It gets fairly awkward without it. We can't strictly look at the last iteration to get the issuer for the current iteration like we do in some other "build a chain" type loops because when we get to the client cert role the last iteration will have been for an end entity cert. We could save the last issuer in separate var only updated for certain roles, but then we also have to do the same for the trust anchor to be able to write a chain file for the EE or client cert role.

In the end since this is a utility only run to generate test data it felt like the approach using the hashmap had the best clarity overall.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+};
+use time::OffsetDateTime;
+
+fn main() {
```
Maybe make this yield `Result<(), rcgen::Error>` so we can `?` all the errors? The try operator is quite a bit less noisy than `.unwrap()`...

(`Result<(), Box<dyn std::error::Error>>` also works if we need to unwrap other errors often.)

cpu replied:
Done, with the exception of `Option` `unwrap()`s where I didn't want to have to invent an error to map the `None` case to.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+    // the end entity or client cert as well.
+    for role in [Role::EndEntity, Role::Client] {
+        for alg in sig_algs {
+            let CertifiedKey { cert: root, .. } = certified_keys
```
Is all the destructuring really worth it? I find these vertically squashed together three-line destructuring expressions kind of hard to read, and not sure they buy us a lot especially in the case where we're only taking a single field.

Compare:

```rust
            let CertifiedKey { cert: root, .. } = certified_keys
                .get(&(Role::TrustAnchor, alg))
                .unwrap();
            let CertifiedKey {
                cert: intermediate, ..
            } = certified_keys
                .get(&(Role::Intermediate, alg))
                .unwrap();
```

To

```rust
let root = certified_keys.get(&(Role::TrustAnchor, alg)).unwrap().cert;
let intermediate = certified_keys.get(&(Role::Intermediate, alg)).unwrap().cert;
```

(Don't know if rustfmt will actually allow these to be on one line...)

cpu replied:
I removed the destructuring. You're right, it was noisey and wasn't providing much value.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+            } = certified_keys
+                .get(&(role, alg))
+                .unwrap();
+            fn write_chain_file(
+                role: Role,
+                alg: &'static SignatureAlgorithm,
+                name: &str,
+                chain_certs: &[&Certificate],
+            ) {
+                let mut chain_file =
+                    File::create(output_directory(alg).join(format!("{}.{}", role.label(), name)))
+                        .unwrap();
+                for cert in chain_certs {
+                    chain_file
+                        .write_all(cert.pem().as_bytes())
+                        .unwrap();
+                }
+            }
+
+            write_chain_file(role, alg, "chain", &[intermediate, root]);
+            write_chain_file(role, alg, "fullchain", &[end_entity, intermediate, root]);
```
Instead of a function, consider a loop over `let files = [("chain", &[intermediate, root]), ("fullchain", &[end_entity, intermediate, root])];"?

cpu replied:
Done.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+    key_pair: KeyPair,
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
```
Nit: maybe move this down.

cpu replied:
Moved down.

djc reviewed (commented):
Turns out I did have some feedback... hope it's useful.

cpu commented:
> cpu marked this pull request as draft now

Putting this back to draft until I have a chance to rebase + address feedback. It's a bit lower on my priority list ATM.

cpu commented on provider-example/examples/server.rs:
```diff
@@ -70,25 +70,21 @@ impl TestPki {
         Self {
-            server_cert_der,
-            server_key_der,
+            server_cert_der: server_cert.der().clone(),
```
Fixed

cpu commented on examples/src/bin/server_acceptor.rs:
```diff
@@ -272,19 +276,19 @@ impl TestPki {
+        crl_params
+            .signed_by(&self.ca_cert.cert, &self.ca_cert.key_pair)
             .unwrap()
+            .der()
```
Fixed.

cpu commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+    } else {
+        KeyPair::generate_for(alg)
+    };
+    key_pair.unwrap()
```
> Skip the key_pair binding and chain the unwrap()?

Done.

> Maybe use a for-loop over [(PKCS_RSA_SHA256, RsaKeySize::_2048), ..]?

This ended up being awkward so I skipped this suggestion.

cpu commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,365 @@
+                params.key_usages = vec![
+                    KeyUsagePurpose::DigitalSignature,
+                    KeyUsagePurpose::ContentCommitment,
+                ];
```
Done.

cpu commented:
> Putting this back to draft until I have a chance to rebase + address feedback.

I've addressed the feedback and this branch is ready for another round of review.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,321 @@
+}
+
+// Note: these are ordered such that the data dependencies for issuance are satisfied.
+const ROLES: [Role; 4] = [
+    Role::TrustAnchor,
+    Role::Intermediate,
+    Role::EndEntity,
+    Role::Client,
+];
```
Nit: I guess this should go before the definition of `Role`.

cpu replied:
Fixed.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,321 @@
+            cert_file.write_all(cert.der())?;
+
+            // If we're not a trust anchor, generate a CRL for the certificate we just issued.
+            if !matches!(role, Role::TrustAnchor) {
```
Nit: since `Role` implements `PartialEq` anyway, just write `role != Role::TrustAnchor`?

cpu replied:
Ah! Good point. Fixed.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,321 @@
+    Ok(())
+}
+
+fn crl_for_serial(serial_number: &SerialNumber) -> CertificateRevocationListParams {
```
Nit: maybe consider taking `SerialNumber` directly here, moving the `clone()` to the outside? The caller can just use `cloned()` instead of `as_ref()` so that's a small difference.

cpu replied:
Good idea, fixed.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,321 @@
+        }
+
+        // Client certificates additionally get the client auth EKU.
+        if matches!(self, Self::Client) {
```
Nit: rely on `PartialEq`?

cpu replied:
Done.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,321 @@
+}
+
+impl Role {
+    fn key_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.key", self.label()))
+    }
+
+    fn cert_pem_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.cert", self.label()))
+    }
+
+    fn cert_der_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.der", self.label()))
+    }
+
+    fn label(&self) -> &'static str {
+        match self {
+            Self::Client => "client",
+            Self::EndEntity => "end",
+            Self::Intermediate => "inter",
+            Self::TrustAnchor => "ca",
+        }
+    }
```
Nit: move all these simple getters below `params()` and `common_name()`?

cpu replied:
Done.

djc commented on rustls/examples/internal/test_ca.rs:
```diff
@@ -0,0 +1,321 @@
+                params.is_ca = IsCa::NoCa;
+                params.key_usages = EE_KEY_USAGES.to_vec();
+                params.subject_alt_names = vec![
+                    SanType::DnsName(Ia5String::try_from("testserver.com".to_string()).unwrap()),
```
Nit: where building a `String` from a `&str`, prefer `to_owned()` over `to_string()` (which is more general and more expensive).

cpu replied:
Done.

djc reviewed (approved):
Much better -- still some small suggestions.
